### ISSUE TITLE
Execution of expressions with variables as operations

### ISSUE THREAD

**Issue by Necr0x0Der** [trueagi-io/hyperon-experimental#242]
The following code:
```
(= (inc $x) (+ $x 1))
(= (dec $x) (- $x 1))
!($x 2)
```
returns `[1, 3]`. This is fun, and doesn't contradict to evaluation via equality queries. However, I'd prefer to
1) be able to specify expressions like `($x $y)` without automatic reduction;
2) force to explicitly match functional symbols if needed, e.g.
```
!(match &self (= ($func 2) $r) $r)
!(match &self (= ($func $x) $r) ($func 2))
```
both work.
Thus, I propose to consider variables as non-operational if this doesn't require considerable refactoring and breaks no tests.

**Comment by ngeiswei** [comment]
I guess both semantics (operational and non-operational variables) have upsides and downsides.  The downside of the operational variable semantics is that it is more prone to infinite recursion.  Its upside is that it is more elegant as far as defining evaluation as a form of matching.

Maybe a pragma could be introduced to switch between these two semantics?

Or, maybe even better, that behavior could be more generally controlled by overwriting some mini-metta code (though I don't know precisely how, if at all possible)?

**Comment by ngeiswei** [comment]
An inelegance of the operational variable semantics, as currently implemented, is that built-in functions are treated differently than metta functions.  For instance `($x $y $z $w)` unifies with `(if ...)` but not with `(let ...)`.

**Comment by vsbogd** [comment]
> An inelegance of the operational variable semantics, as currently implemented, is that built-in functions are treated differently than metta functions. For instance ($x $y $z $w) unifies with (if ...) but not with (let ...).

It is more inelegance of the grounded functions implementation. As grounded functions are not parts of the atomspace, they cannot be matched by `($x ...)` patterns. It is one of the things we can fix by fixing grounded functions representation.

**Comment by vsbogd** [comment]
> Or, maybe even better, that behavior could be more generally controlled by overwriting some mini-metta code (though I don't know precisely how, if at all possible)?

Yes, one of the ways to make it flexible is allow to redefine `eval` function to the user. One of the ideas is to provide a way to implement `eval` via `match`.

**Comment by noskill** [comment]
why it is blocker? What it blocks? @Necr0x0Der

**Comment by vsbogd** [comment]
> why it is blocker? What it blocks? @Necr0x0Der

Anatoly it is more a question to @ngeiswei . Basically the part of PLN code generates templates like `($x $y $z)` to be matched with space and they are executed by interpreter as they can be matched with `(= (<op> <arg1> <arg2>) <anything>)` templates in an atomspace.

**Comment by noskill** [comment]
It seems to be very common request when coding in metta: to pass some atoms as they are without evaluation. Perhaps it can be solved with types?

**Comment by noskill** [comment]
This program stil matches operations and executes them:

        (: makelist (-> Atom Atom))
        (= (makelist $A)
            (if (== () $A) Nil (Cons (car-atom $A) (makelist (cdr-atom $A)))))

        !(makelist ($A $B $C))
        

i build with this in Cargo.toml

[features]
#default = ["variable_operation"]
default = ["variable_operation", "minimal"]
minimal = []
variable_operation = []

@ngeiswei @luketpeterson

**Comment by noskill** [comment]
I expected Atom type to disable execution.. should it work this way in minimal metta?

**Comment by vsbogd** [comment]
> I expected Atom type to disable execution.. should it work this way in minimal metta?
Yes

**Comment by vsbogd** [comment]
@noskill  to not make a confusion if you see the issue with minimal MeTTa interpreter let raise it as a separate issue as minimal MeTTa is still under construction. I will add a label `minimal` for such kind of issues.

**Comment by Necr0x0Der** [comment]
The issue is that you need to remove `"variable_operation"`, not add it. See https://github.com/trueagi-io/hyperon-experimental/issues/488#issuecomment-1798055998 for a working code example (note that `(makelist (cdr-atom $A))` will not reduce `(cdr-atom $A)` before calling `makelist` with `Atom` argument type.

**Comment by Necr0x0Der** [comment]
With `"variable_operation"` removed, 
```(= (inc $x) (+ $x 1))
(= (dec $x) (- $x 1))
!($x 2)```
is not reduced, while
`!(match &self (= ($func 2) $r) $r)` produces `[1, 3]`.

I would actually remove `"variable_operation"` as default behavior, because if someone wants to match against functions, they can use `match`, while people are being confused by evaluation of `($A $B)` by default.
Any opinion?

**Comment by Necr0x0Der** [comment]
I'm leaning to close this issue since it is solved. I hesitate a little, because the default behavior remains confusing, but maybe it doesn't matter much since we are going to migrate to Minimal MeTTa. @vsbogd you can either close this issue or remove `"variable_operation"` from defaults and close it after this. It's up to you.

**Comment by noskill** [comment]
@vsbogd I expect !(cdr-atom ($A $B $C)) to return ($B $C), trying it on minimal metta

**Comment by vsbogd** [comment]
`"variable_operation"` is left in defaults because it was discussed with @ngeiswei in PR (see https://github.com/trueagi-io/hyperon-experimental/pull/473#issuecomment-1782627167)

**Comment by vsbogd** [comment]
I am closing it for the issue posted by @noskill above we have separate https://github.com/trueagi-io/hyperon-experimental/issues/488

**Comment by Necr0x0Der** [comment]
@vsbogd 
I would still prefer `"variable_operation"` to be not enabled by default. See my the first and last messages

**Comment by Necr0x0Der** [comment]
My main point is that, when `"variable_operation"` is turned off, one can achieve both kinds of behavior with using `match`. But when it is turned on, only one kind of behavior is available.

**Comment by Necr0x0Der** [comment]
I'd like to hear the argument against this consideration. As I see from the discussion

> I don't mind either way, as long as the flag is mentioned in the README.md. Maybe old behavior could be the default to minimize disruption for now.

So, I wonder "for now" or "forever"? If it is a temporary choice to have the old behavior as default, we need to keep this issue open to solve it later, if not now.

**Comment by Necr0x0Der** [comment]
If the old behavior is supposed to be kept forever, let's decide on this explicitly.

**Comment by vsbogd** [comment]
Yeah, I think that fix we have is a temporary fix and more proper fix should include some mechanism to control this behavior from the program itself. Thus I agree with issue reopening. But I think we can remove `blocker

**Comment by Necr0x0Der** [comment]
> to control this behavior from the program itself

I'm actually satisfied with the current solution. As I said, we can always call `match` explicitly if needed `!(match &self (= ($func 2) $r) $r)`, so it is kind of controllable by the program itself. But, yes, `#pragma` might be nicer than a compilation feature. I'd also like to test the behavior after migration to minimal metta happens, so we can keep it open for a while.

**Comment by ngeiswei** [comment]
There are subtle situations where the function itself is an expression (as opposed to just a symbol).  For instance

```
(= (((λ $x) $x) $y) $y)
!(($l $x) $y)
```

outputs

```
[$y]
```

Meaning that `($l $x)` unifies with `((λ $x) $x)` and triggers the reduction, even when MeTTa is built with `variable_operation` disabled.

It may seem like a desirable behavior but it may result in undesirable outcomes in a situation such as

```
(= (((λ $x) $f) $y) (let $x $y $f))
!(((λ $x) $x) ((λ $f) ((λ $g) ((λ $z) (($f $z) ($g $z))))))
```

which outputs

```
[((λ (λ $x#3186)) ((λ $g) ((λ $z) $z)))]
```

instead of (something alpha equivalent to)

```
[((λ $n) ((λ $f) ((λ $x) ($f (($n $f) $x)))))]
```

The reason it seems is because `(($f $z) ($g $z))` then unifies to `((λ $x) $f) $y)` and triggers reduction.

**Comment by ngeiswei** [comment]
```
(= (((λ $x) $x) $y) $y)
!((λ $f) ((λ $g) ((λ $z) (($f $z) ($g $z)))))
```

outputs

```
((λ (λ $x#2371)) ((λ $g) ((λ $z) ($g $z))))
```

which I believe is explained by the fact that `(($f $z) ($g $z))` indeed unifies with `((λ $x) $x) $y)` therefore substituting `(($f $z) ($g $z))` by `($g $z)` as well as substituting `(λ $f)` by `(λ (λ $x#2371))`.

**Comment by ngeiswei** [comment]
Please find a shorter code with a similar behavior

```
(= (((λ $x) $x) $y) $y)
!((λ $f) (($f $z) $g))
```

outputs

```
[((λ (λ $x#1010)) $g)]
```

**Comment by Adam-Vandervorst** [comment]
Following [Linear Unification](https://www.sciencedirect.com/science/article/pii/0022000078900430) Algorithm C:

```
(= (((λ $x) $x) $y) $y)
!(($l $x) $y)
```
Finishes with a failed unification `pointer((1 ◆)) = (1 ◆) instead of null` (would be an infinite term)

```
(= (((λ $x) $x) $y) $y)
!((λ $f) (($f $z) $g))
```
Finishes with a failed unification `(λ ◆) and (λ ◆ ⏴₁) do not match` (no MGU)

**Comment by vsbogd** [comment]
```
(= (((λ $x) $x) $y) $y)
!((λ $f) (($f $z) $g))
```
This example is reduced partially. `(($f $z) $g)` is unified with `(((λ $x) $x) $y)` and result `{ $X = $g, $z = $x#9, $f <- (λ $x#9) }` is applied to the whole expression. Thus final result is `((λ (λ $x#9)) $g)`.

**Comment by ngeiswei** [comment]
> Thus final result is `((λ (λ $x#9)) $g)`

Yes, that's my problem, it does not behave like regular lambda calculus.  My general question would be how to make it behave like lambda calculus.

**Comment by Adam-Vandervorst** [comment]
What does "reduced partially" mean here? If it's not the most general unifier (isn't finite), what is returned?

**Comment by vsbogd** [comment]
#668 is also related to this issue

**Comment by vsbogd** [comment]
Should we close it?

**Comment by ngeiswei** [comment]
As fare as I am concerned it can be closed, the behavior is now correct according to my tests.
